### ISSUE TITLE
Ordering of `--network` flags is disregarded

### ISSUE THREAD

**Issue by dpward** [containers/podman#12850]
/kind bug

When creating a container, multiple CNI networks can be connected to it, using this syntax (or alternatives):
```
podman container create --network=cni-podman2 --network=cni-podman3 --network=cni-podman1
```

However, the ordering is disregarded when [parsing the network flags](https://github.com/containers/podman/blob/v4.0.0-rc1/pkg/specgen/namespaces.go#L330-L445). This happens because it returns the network names as keys in an unordered map. So the networks are always added to the container in alphabetical order, i.e.
- `eth0` will be created for `cni-podman1`
- `eth1` will be created for `cni-podman2`
- `eth2` will be created for `cni-podman3`

The expected behavior is that
- `eth0` will be created for `cni-podman2`
- `eth1` will be created for `cni-podman3`
- `eth2` will be created for `cni-podman1`

The most obvious way to accomplish this is to add a fourth return value from `ParseNetworkFlags()`, which is a slice containing the network names in the order they are parsed. (However, there may be other solutions.)

cc: @Luap99

**Comment by Luap99** [comment]
Is that actually an issue, there was never a guarantee that the interface name is consistent with the ordering of the network names. That this worked before is more of an implementation detail than a feature.

I don't think we should fix this but if you really need predictable interface names you can use `--network net1:interface_name=eth0` and so on.

**Comment by Luap99** [comment]
You can already set the interface name in the `types.PerNetworkOptions` struct so you do not need to alter the function signature at all. But I still do not think we want this. What is your actual use case that would require this?

**Comment by dpward** [comment]
You can do that - but still if you don't, we would expect the interface ordering on the command line to be preserved. This affects the ifindex as well. (We are running routing software in containers.)

Could you please be a bit more clear about why this is a bad idea, rather than something that simply is not implemented?

**Comment by Luap99** [comment]
It's not a bad idea I just don't think that guaranteeing this is worth the extra maintenance in the long run. The naming thing can be fixed easily and we could add a test to prevent regressions but this will not fix the ifindex. In the backend we use maps so if we loop over the map the ordering is not deterministic.

**Comment by dpward** [comment]
It would be very useful to me. I've examined the backend implementation. All we would need to do is to have a slice containing the interface names in the correct order, and iterate over that to obtain the values from the map in the order we want.

Does that sound reasonable if I prepare a PR?

**Comment by Luap99** [comment]
Are you talking about only the naming or also the ifindex?

If you need the stable ifindex, what would be the correct order, interface names sorted ascending?
I am not sure were you looked to implement this but if you do so you have to do it for both cni and netavark.
What happens with podman network connect/disconnect?
I still do not think the added complexity is worth the extra maintenance.

**Comment by Luap99** [comment]
@mheon WDYT?

**Comment by dpward** [comment]
I'm talking about just adding the interfaces in the order they appear on the command line. As a result the first interface will automatically have a lower ifindex than the second, etc. Note that commands like `ip link` show the interfaces in ifindex order.

**Comment by mheon** [comment]
Would the ability to manually set interface names (e.g. `--net cni-podman1:ifname=eth0 --net cni-podman2:ifname=eth1 ...`) be acceptable? I'm a little iffy on making ordering in the CLI a critical part of the interface, but I definitely see how consistent interface naming is useful

**Comment by dpward** [comment]
Right now we have no way to control the ifindex ordering. It depends on when the interface is created, not what name we reassign it. So creating them in the order they appear on the command line would matter.

To give just one example of why this can be useful, the `ip link` command displays the interfaces in ifindex order - and if the container has dozens of interfaces, that matters a lot more. (The exception to this is `host-device` networks which preserve the existing ifindex from the host.)

**Comment by Luap99** [comment]
I am not sure how you want to fix this but it is not possible to keep the cli order in the backend since it is stored as a map. You could sort by the interface name but this would still not match the cli ordering if I would use `--network net1:interface_name=eth1 --network net2:interface_name=eth0`.

Just the naming change is simple if that is enough for you.

**Comment by mheon** [comment]
I suppose we could automatically add `interface_name` in order when we first parse into the maps (if it's not explicitly set)

**Comment by Luap99** [comment]
@mheon Yes that is easy but he also wants the ifindex to be in the same order and this depends on the order in which the interfaces are created

**Comment by mheon** [comment]
Ahh. Yeah, that would be more difficult.

**Comment by dpward** [comment]
In the [libpod API](https://docs.podman.io/en/latest/_static/api.html#operation/ContainerCreateLibpod), we already have the `cni_networks` attribute, which is an ordered list of network names to be used in the container. If we specify that, then we can add the networks order.

The documentation says that the new `Networks` attribute is supposed to be used in favor of that, although `cni_networks` remains for compatibility. I'm actually a bit surprised this `Networks` doesn't deprecate `network_options`: these are both maps, where the key is the network name, and the value contains options specific to that network. (I realize I haven't been following the development discussions here.)

**Comment by Luap99** [comment]
The API level doesn't matter here, we are talking about the backend. It is stored as map in the db so there is no way to keep the order. We would need to change the types and db to make this work which is not worth the effort also podman 4.0 is about to be released very soon so there is no time to change this.

`cni_networks` is for backwards compatibility only
`network_options` is only used for slirp4netns at the moment, I could have used that for the networks and there is no particular reason why I didn't (maybe that would have made more sense in hindsight I don't know)

Note that just looking at `ip link` output is not an argument for me. We can make the interface naming change but the ifindex things is not a simple as you think.

**Comment by dpward** [comment]
I realize it's a bit more involved than at first look, but I'd still like to figure out what a workable solution would be here.

To provide another example, let's say I have two `host-device` networks. If attaching the first one fails, then podman won't attempt to attach the second. When these networks are returned to the host, they don't have the same configuration they had originally. Having determinism in the order that interfaces are added to the container would again be useful here.

**Comment by daiaji** [comment]
This should be helpful for using soft routing in containers.

**Comment by rhatdan** [comment]
@Luap99 Is this still an issue?

**Comment by Luap99** [comment]
Yes

**Comment by rhatdan** [comment]
@Luap99 What do you think about doing something like the following patch and retaining the order that the user added the objects.

[order.patch.txt](https://github.com/containers/podman/files/8350021/order.patch.txt)

**Comment by Luap99** [comment]
I am not a huge fan of adding a new field for this, this will everything more complicated, this can be set by api users so we would need to check for invalid user input.
IMO if we want to make it deterministic we should sort by the interface name, then we can apply the names in the frontent in the order which --network was given. But adding the names in the frontent would also conflict with https://github.com/containers/podman/issues/13174

**Comment by rhatdan** [comment]
Let discuss this on Monday with @mheon at standup.

**Comment by rhatdan** [comment]
@Luap99 @mheon We never followed up on this.  What do you want to do with this one?

**Comment by mheon** [comment]
I'm OK with fixing but IMO this is low priority. Exact way to implement still requires some discussion.

**Comment by rhatdan** [comment]
Since this was low priority and no one ever acted on it, I am going to close.  Reopen if you want to work on it.

**Comment by Luap99** [comment]
I agree that is a valid feature request, it is just that work for it seems rather high.
If you have support agreements then please use them.

**Comment by livelace** [comment]
it affects my setup, i need specific network interfaces be placed into specific bridges for multus' proper network mapping. i was confident that `--network` arg is ordered.

**Comment by ashb** [comment]
A case where this has just caused me problems.

I have defined a network using macvlan driver with the dhcp IPAM plugin/driver, and when I have a multi-homed container (one network is macvlan, the other is a "normal" bridge internal network) _sometimes_ the pod-to-pod DNS works, other times it doesnt.

"working":

```
# cat /etc/resolv.conf
search dns.podman
nameserver 10.89.1.1
nameserver 192.168.152.17
nameserver 192.168.152.3
```

"broken":

```
# cat /etc/resolv.conf
search dns.podman
nameserver 192.168.152.17
nameserver 192.168.152.3
nameserver 10.89.1.1
```

**Comment by Luap99** [comment]
This is planned for 6.0, https://github.com/containers/podman/blob/main/contrib/design-docs/podman6hld.md#preserve-network-order
internal jira card https://issues.redhat.com/browse/RUN-3573 to track progress

I already patched netavark to allow ordered setup in https://github.com/containers/netavark/pull/1369
